core/libp2p/muxer/yamux/yamux.cpp

```diff
@@ -32,20 +32,18 @@ namespace libp2p::muxer {
     }
 
     // see, if there's any messages to be written to the connection
-    {
-      std::lock_guard<std::mutex> lg{outcoming_msgs_mutex_};
-      if (!outcoming_messages_.empty()) {
-        auto msg_and_callback = outcoming_messages_.front();
-        outcoming_messages_.pop();
-        connection_->writeAsync(
-            msg_and_callback.first,
-            [t = shared_from_this(), cb = std::move(msg_and_callback.second)](
-                const boost::system::error_code &ec, size_t n) {
-              t->writingComplete(ec, n, std::move(cb));
-            });
-      }
+    while (!outcoming_messages_.empty()) {
+      auto msg_and_callback = outcoming_messages_.front();
+      outcoming_messages_.pop();
+      connection_->writeAsync(
+          msg_and_callback.first,
```

Comment:
msg_and_callback will be destroyed on the next iteration, writeAsync will fail with UB.
you, as caller, MUST ensure that buffer is valid until completion handler is executed